File: mac/Voice/Refine/RefineService.swift
import Foundation
import os

private let log = Logger(subsystem: "co.miraside.voice", category: "refine")

/// Budgeted refine with raw fallback; outbox replay after any successful call.
/// G5: main-actor — the `Refiner` protocol's async methods are called from the @MainActor Coordinator.
@MainActor
final class RefineService: Refiner {
    private let api: VoiceAPIClient
    private let outbox: Outbox
    private var pendingByClientId: [UUID: (Dictation, FallbackReason?)] = [:]
    var asrModel = Preferences.modelId   // C1: Preferences is a static namespace

    init(api: VoiceAPIClient, outbox: Outbox) { self.api = api; self.outbox = outbox }

    func refine(_ d: Dictation, mode: String) async -> RefineResult {
        guard let raw = d.raw else { return .rawFallback(.server) }
        let ctx = RefineRequest.Context(appBundleId: d.app?.bundleId, appName: d.app?.name)
        let timing = RefineRequest.Timing(audioMs: d.audioMs, asrMs: d.asrMs)
        let createdAt = Int(d.startedAt.timeIntervalSince1970 * 1000)
        let languageSetting = Preferences.language

        if mode == "literal" {
            await logOnly(d, injected: .raw, fallback: nil, mode: mode)
            return .literal
        }
        let budget = Budget.ms(rawChars: raw.count)
        let body = RefineRequest(clientId: d.clientId.uuidString.lowercased(), raw: raw, mode: mode, languageSetting: languageSetting, languageDetected: d.language,
                                 budgetMs: budget, context: ctx, timing: timing, asrModel: asrModel, clientVersion: VoiceAPI.version, createdAt: createdAt)
        do {
            let r = try await api.refine(body, budgetMs: budget)
            await replayOutbox()
            if let reason = r.fallbackReason { return .rawFallback(reason) }
            return r.cleaned.isEmpty ? .rawFallback(.guardRejected) : .cleaned(r.cleaned)
        } catch let e as APIError {
            let reason: FallbackReason = switch e {
            case .timeout: .clientTimeout
            case .offline: .offline
            case .unauthorized: .unauthorized
            case .server, .decoding: .server
            }
            pendingByClientId[d.clientId] = (d, reason)
            log.info("refine fallback \(reason.rawValue, privacy: .public)")
            return .rawFallback(reason)
        } catch {
            pendingByClientId[d.clientId] = (d, .server)
            return .rawFallback(.server)
        }
    }

    func reportInjected(clientId: UUID, injected: Injected) async {
        if let (d, reason) = pendingByClientId.removeValue(forKey: clientId) {
            // The refine never completed on our side: log it (or queue it) instead of patching.
            await logOnly(d, injected: injected, fallback: reason, mode: "clean")
            return
        }
        // C3: PATCH 404s until a refine/dictation POST has stored the row for this clientId — this
        // only runs after a successful refine (the branch above handles the "never stored" case), so a
        // 404 here means the earlier POST itself failed; log it, do not retry.
        // G6: the error kind is useful (offline vs. 404 vs. decode failure); APIError carries no
        // transcript text, so it's fine as .public. clientId is a UUID, also fine as .public.
        do { try await api.patchInjected(clientId: clientId, injected: injected) } catch { log.info("patch injected failed: \(String(describing: error), privacy: .public)") }
    }

    private func logOnly(_ d: Dictation, injected: Injected, fallback: FallbackReason?, mode: String) async {
        // G2: pass `fallback` through as-is — literal mode calls this with `nil`, and that must reach
        // the server as `fallbackReason: null`, not get coalesced into a false ".offline".
        let entry = OutboxEntry(clientId: d.clientId, raw: d.raw ?? "", injected: injected, fallback: fallback, createdAt: d.startedAt,
                                mode: mode, languageSetting: Preferences.language, languageDetected: d.language, app: d.app, audioMs: d.audioMs, asrMs: d.asrMs)
        do { try await api.postDictation(request(for: entry)) } catch { outbox.add(entry) }
    }

    private func replayOutbox() async {
        let entries = outbox.drain()
        var failed: [OutboxEntry] = []
        for e in entries { do { try await api.postDictation(request(for: e)) } catch { failed.append(e) } }
        outbox.requeue(failed)
    }

    private func request(for e: OutboxEntry) -> DictationRequest {
        DictationRequest(clientId: e.clientId.uuidString.lowercased(), raw: e.raw, injected: e.injected, fallbackReason: e.fallback, mode: e.mode,
                         languageSetting: e.languageSetting, languageDetected: e.languageDetected,
                         context: .init(appBundleId: e.app?.bundleId, appName: e.app?.name), timing: .init(audioMs: e.audioMs, asrMs: e.asrMs),
                         asrModel: asrModel, clientVersion: VoiceAPI.version, createdAt: Int(e.createdAt.timeIntervalSince1970 * 1000))
    }
}

File: windows/Spit.Core/Refine/RefineService.cs
using System.Collections.Concurrent;
using System.Globalization;

namespace Spit.Core;

/// Budgeted refine with raw fallback; outbox replay after any successful call. Port of
/// mac/Voice/Refine/RefineService.swift. The Mac's `Preferences.language`, `Preferences.modelId` and
/// `VoiceAPI.version` globals are injected. Nothing here logs or persists transcript text (rule 25).
public sealed class RefineService : IRefiner
{
    private readonly IVoiceApiClient _api;
    private readonly Outbox _outbox;
    private readonly ILocalSettings _settings;
    private readonly string _clientVersion;

    /// Dictations whose refine never completed on our side, with the reason it failed. Reported through
    /// `POST /v1/dictations` (or the outbox) instead of a PATCH, since the server has no row to patch.
    private readonly ConcurrentDictionary<Guid, (Dictation Dictation, FallbackReason? Reason)> _pendingByClientId = new();

    public RefineService(IVoiceApiClient api, Outbox outbox, ILocalSettings settings, string asrModel, string clientVersion)
    {
        _api = api;
        _outbox = outbox;
        _settings = settings;
        AsrModel = asrModel;
        _clientVersion = clientVersion;
    }

    /// Sent as `asrModel`; settable because the user can switch models without relaunching.
    public string AsrModel { get; set; }

    public async Task<RefineResult> RefineAsync(Dictation dictation, string mode)
    {
        var d = dictation;
        if (d.Raw is not { } raw) return new RefineResult.RawFallback(FallbackReason.Server);
        var ctx = new RefineContext(d.App?.BundleId, d.App?.Name);
        var timing = new RefineTiming(d.AudioMs, d.AsrMs);
        var createdAt = d.StartedAt.ToUnixTimeMilliseconds();
        var languageSetting = _settings.Language;

        if (mode == "literal")
        {
            await LogOnlyAsync(d, Injected.Raw, fallback: null, mode).ConfigureAwait(false);
            return new RefineResult.Literal();
        }

        // Rules 11-14: a transcript that needs nothing skips the round trip entirely. Still logged, with
        // `llmModel` marking it, because the skip rate is the only signal that says whether the gate's
        // thresholds are right.
        if (SkipGate.ReasonToClean(raw, mode, d.Language, d.StreamedSegments) is null)
        {
            await LogOnlyAsync(d, Injected.Raw, fallback: null, mode, llmModel: CleanupEngine.Skipped).ConfigureAwait(false);
            return new RefineResult.Skipped();
        }

        var budget = Budget.Ms(new StringInfo(raw).LengthInTextElements);
        var body = new RefineRequest(d.WireId, raw, mode, languageSetting, d.Language, budget, ctx, timing,
            AsrModel, _clientVersion, createdAt);
        try
        {
            var r = await _api.RefineAsync(body, budget).ConfigureAwait(false);
            await ReplayOutboxAsync().ConfigureAwait(false);
            if (r.FallbackReason is { } serverReason) return new RefineResult.RawFallback(serverReason);
            return string.IsNullOrEmpty(r.Cleaned)
                ? new RefineResult.RawFallback(FallbackReason.GuardRejected)
                : new RefineResult.Cleaned(r.Cleaned);
        }
        catch (ApiException e)
        {
            var reason = e.Kind switch
            {
                ApiErrorKind.Timeout => FallbackReason.ClientTimeout,
                ApiErrorKind.Offline => FallbackReason.Offline,
                ApiErrorKind.Unauthorized => FallbackReason.Unauthorized,
                _ => FallbackReason.Server,   // server(status) and decoding
            };
            _pendingByClientId[d.ClientId] = (d, reason);
            return new RefineResult.RawFallback(reason);
        }
        catch (Exception)
        {
            _pendingByClientId[d.ClientId] = (d, FallbackReason.Server);
            return new RefineResult.RawFallback(FallbackReason.Server);
        }
    }

    public async Task ReportInjectedAsync(Guid clientId, Injected injected, int? totalMs)
    {
        if (_pendingByClientId.TryRemove(clientId, out var pending))
        {
            // The refine never completed on our side: log it (or queue it) instead of patching.
            await LogOnlyAsync(pending.Dictation, injected, pending.Reason, "clean", totalMs: totalMs).ConfigureAwait(false);
            return;
        }
        // Mac C3: the PATCH 404s until a POST has stored the row. This only runs after a successful
        // refine, so a failure here means that POST itself failed — do not retry.
        try
        {
            await _api.PatchInjectedAsync(clientId, injected, totalMs).ConfigureAwait(false);
        }
        catch (Exception)
        {
            // Nothing to surface: the dictation was already pasted.
        }
    }

    private async Task LogOnlyAsync(Dictation d, Injected injected, FallbackReason? fallback, string mode,
        string? llmModel = null, int? totalMs = null)
    {
        // Mac G2: `fallback` passes through as-is — literal mode sends null, which must reach the server as
        // no fallback, not get coalesced into a false `offline`.
        var entry = new OutboxEntry(d.ClientId, d.Raw ?? "", injected, fallback, d.StartedAt, mode, _settings.Language,
            d.Language, d.App, d.AudioMs, d.AsrMs)
        {
            LlmModel = llmModel,
            TotalMs = totalMs,
        };
        try
        {
            await _api.PostDictationAsync(Request(entry)).ConfigureAwait(false);
        }
        catch (Exception)
        {
            _outbox.Add(entry);
        }
    }

    private async Task ReplayOutboxAsync()
    {
        var entries = _outbox.Drain();
        var failed = new List<OutboxEntry>();
        foreach (var e in entries)
        {
            try
            {
                await _api.PostDictationAsync(Request(e)).ConfigureAwait(false);
            }
            catch (Exception)
            {
                failed.Add(e);
            }
        }
        _outbox.Requeue(failed);
    }

    private DictationRequest Request(OutboxEntry e) =>
        new(e.ClientId.ToString("D"), e.Raw, e.Injected, e.Fallback, e.Mode, e.LanguageSetting, e.LanguageDetected,
            new RefineContext(e.App?.BundleId, e.App?.Name), new RefineTiming(e.AudioMs, e.AsrMs),
            AsrModel, _clientVersion, e.CreatedAt.ToUnixTimeMilliseconds())
        {
            Cleaned = e.Cleaned,
            LlmMs = e.LlmMs,
            LlmModel = e.LlmModel,
            TotalMs = e.TotalMs,
        };
}
